packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -143,203 +192,138 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
-            abi.encodePacked(_vaultId, _wPowerPerpAmountToBurn, _wPowerPerpAmountToBuy, _collateralToWithdraw)
+            _params.wPowerPerpAmountToBurn.add(_params.wPowerPerpAmountToBuy),
+            _params.maxToPay,
+            uint8(CALLBACK_SOURCE.FLASH_W_BURN),
+            abi.encode(_params)
         );
 
-        emit FlashWBurn(msg.sender, _vaultId, _wPowerPerpAmountToBurn, _collateralToWithdraw, _wPowerPerpAmountToBuy);
+        payable(msg.sender).sendValue(address(this).balance);
+
+        emit FlashWBurn(
+            msg.sender,
+            _params.vaultId,
+            _params.wPowerPerpAmountToBurn,
+            _params.collateralToWithdraw,
+            _params.wPowerPerpAmountToBuy
+        );
     }
 
     /**
      * @notice sell long wPowerPerp and flashswap mint short position
-     * @dev flahswap amount = collateral amount - msg.value - ETH from selling long wPowerPerp
-     * @param _vaultId vault ID
-     * @param _wPowerPerpAmountToMint wPowerPerp amount to mint
-     * @param _collateralAmount collateral amount to use for minting
-     * @param _wPowerPerpAmountToSell long wPowerPerp amount to sell
-     * @param _minToReceive min ETH amount to receive for selling long _wPowerPerpAmountToMint+_wPowerPerpAmountToSell
+     * @dev flashswap amount = collateral amount - msg.value - ETH from selling long wPowerPerp
+     * @param _params FlashSellLongWMintParams struct
      */
-    function flashswapSellLongWMint(
-        uint256 _vaultId,
-        uint256 _wPowerPerpAmountToMint,
-        uint256 _collateralAmount,
-        uint256 _wPowerPerpAmountToSell,
-        uint256 _minToReceive
-    ) external payable {
-        IWPowerPerp(wPowerPerp).transferFrom(msg.sender, address(this), _wPowerPerpAmountToSell);
+    function flashswapSellLongWMint(FlashSellLongWMintParams calldata _params) external payable {
+        IWPowerPerp(wPowerPerp).transferFrom(msg.sender, address(this), _params.wPowerPerpAmountToSell);
 
-        // flahswap and mint short position
+        // flashswap and mint short position
         _exactInFlashSwap(
             wPowerPerp,
             weth,
             IUniswapV3Pool(wPowerPerpPool).fee(),
-            _wPowerPerpAmountToMint.add(_wPowerPerpAmountToSell),
-            _minToReceive,
-            uint8(FLASH_SOURCE.FLASH_SELL_LONG_W_MINT),
-            abi.encodePacked(_vaultId, _wPowerPerpAmountToMint, _collateralAmount)
+            _params.wPowerPerpAmountToMint.add(_params.wPowerPerpAmountToSell),
+            _params.minToReceive,
+            uint8(CALLBACK_SOURCE.FLASH_SELL_LONG_W_MINT),
+            abi.encode(_params)
         );
+
+        payable(msg.sender).sendValue(address(this).balance);
+    }
+
+    /**
+     * @notice close short position with user Uniswap v3 LP NFT
+     * @dev user should approve this contract for Uni NFT transfer
+     */
+    function closeShortWithUserNft(CloseShortWithUserNftParams calldata _params) external {
```

Comment:
lets have a naming PR maybe - this maybe be closeVaultWithUserNft or closeMintWithUserNft?